ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/InMemoryActiveLedgerState.scala

```diff
@@ -0,0 +1,145 @@
+    keys: Map[GlobalKey, AbsoluteContractId],
+    parties: Map[Party, PartyDetails])
+    extends ActiveLedgerState[InMemoryActiveLedgerState] {
+
+  override def lookupActiveContract(cid: AbsoluteContractId): Option[ActiveContract] =
+    activeContracts.get(cid)
+
+  override def lookupContract(cid: AbsoluteContractId): Option[Contract] =
+    activeContracts.get(cid).orElse[Contract](divulgedContracts.get(cid))
+
+  override def keyExists(key: GlobalKey) = keys.contains(key)
+
+  /**
+    * Updates divulgence information on the given active contract with information
+    * from the already existing divulged contract.
+    */
+  private def copyDivulgences(ac: ActiveContract, dc: DivulgedContract): ActiveContract =
+    ac.copy(divulgences = ac.divulgences.unionWith(dc.divulgences)((l, _) => l))
+
+  override def addContract(
+      c: ActiveContract,
+      keyO: Option[GlobalKey]): InMemoryActiveLedgerState = {
+    val newKeys = keyO match {
+      case None => keys
+      case Some(key) => keys + (key -> c.id)
+    }
+    divulgedContracts.get(c.id) match {
+      case None =>
+        copy(
+          activeContracts = activeContracts + (c.id -> c),
+          keys = newKeys
+        )
+      case Some(dc) =>
+        copy(
+          activeContracts = activeContracts + (c.id -> copyDivulgences(c, dc)),
+          divulgedContracts = divulgedContracts - c.id,
+          keys = newKeys
+        )
+    }
+  }
+
+  override def removeContract(cid: AbsoluteContractId, keyO: Option[GlobalKey]) = keyO match {
+    case None => copy(activeContracts = activeContracts - cid)
+    case Some(key) => copy(activeContracts = activeContracts - cid, keys = keys - key)
+  }
+
+  override def addParties(newParties: Set[Party]): InMemoryActiveLedgerState =
+    copy(parties = newParties.map(p => p -> PartyDetails(p, None, true)).toMap ++ parties)
+
+  override def divulgeAlreadyCommittedContract(
+      transactionId: TransactionIdString,
+      global: Relation[AbsoluteContractId, Party],
+      referencedContracts: List[(Value.AbsoluteContractId, AbsoluteContractInst)])
+    : InMemoryActiveLedgerState =
+    if (global.nonEmpty) {
+      val referencedContractsM = referencedContracts.toMap
+      // Note: each entry in `global` can refer to either:
+      // - a known active contract, in which case its divulgence info is updated
+      // - a divulged contract, in which case its divulgence info is updated
+      // - an unknown contract, in which case a new divulged contract is stored
+      val updatedAcs = activeContracts.intersectWith(global) { (ac, parties) =>
+        ac copy (divulgences = ac.divulgeTo(parties, transactionId))
+      }
+      val updatedDcs = divulgedContracts.intersectWith(global) { (dc, parties) =>
+        dc copy (divulgences = dc.divulgeTo(parties, transactionId))
+      }
+      val newDcs = global.foldLeft(Map.empty[AbsoluteContractId, DivulgedContract])(
+        (m, e) =>
```

Comment:
Please use the partial function syntax to pattern match on `e` and give `_1` and `_2` meaningful names.